### PR TITLE
Add defensive check to prevent Mojito gets loading multi times

### PR DESCRIPTION
**Pull request by allmywant**

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -163,6 +163,31 @@ function build()
                 }
             });
             
+            callback(null, file);
+        }))
+        .pipe(through.obj(function(file, enc, callback)
+        {
+            // add defensive check to prevent Mojito gets loading multi times
+            if (!config.allowMultiInstance) {
+                let content = fs.readFileSync(`dist/assets/js/${containerName}.js`, 'utf8');
```
Comment (dapperdrop) [inline gulpfile.js]: Cheers @allmywant - this is a cool way of dealing with the issue! 

As there is some repetition in this block, would it be possible to refactor the code and wrap the guts of it in another function? E.g.
```
function insertDefensiveCheck(filename) {
    let content = fs.readFileSync(filename, 'utf8');
    let position = content.indexOf('Mojito =');
    if (position === -1) {
        position = content.indexOf('Mojito=');
    }
    if (position !== -1) {
        content = content.substr(0, position) + 'if (!window.Mojito || !Mojito.testObjects || !Object.keys(Mojito.testObjects).length) {' + content.substring(position) + '}';
        fs.writeFileSync(filename, content);
    }
}
```

```
// Inside the build function, replace the repetitive code block with:
if (!config.allowMultiInstance) {
    insertDefensiveCheck(`dist/assets/js/${containerName}.js`);
    insertDefensiveCheck(`dist/assets/js/${containerName}.pretty.js`);
}
```

Comment (allmywant) [reply gulpfile.js]: @dapperdrop 
Good idea, I've wrapped the code into a `insertPreventDuplicateLoadingCheck` function.

**Review by dapperdrop — APPROVED** [review]

File: gulpfile.js [hunk 2]
```diff
@@ -60,6 +60,17 @@ function test()
     return runTests();
 }
 
+function insertPreventDuplicateLoadingCheck(fileName) {
```
Comment (kingo55) [inline gulpfile.js]: @dapperdrop @allmywant - minor issue, but I wonder if there's a clearer way to label this function. Like `insertDuplicateContainerCheck`.

File: config.js [hunk 3]
```diff
@@ -0,0 +1,10 @@
+module.exports = {
+    // Name of the container file excluding the file extension (e.g. "name" yields "name.js" when published)
+    containerName: 'mojito',
+    // Development S3 container
+    s3BucketDev: 'mojito-js/jsdev',
+    // Production S3 container
+    s3BucketPRD: 'mojito-js/js',
+    // S3 Region
+    s3Region: 'ap-southeast-2'
+};
```
Comment (kingo55) [inline config.js:12]: Do we need to show an example of `config.allowMultiInstance` setting in this file and in the example file too?

File: tests/test_suite.html [hunk 4]
```diff
@@ -927,6 +927,24 @@
...
+            po.type = 'text/javascript';
+            po.async = false;
+
+            var stags = document.getElementsByTagName('script');
+            var s = document.getElementsByTagName('script')[stags.length -1];
+            s.parentNode.insertBefore(po, s);
+            po.src = '../dist/assets/js/mojito.js';
+            po.onload = function () {
+                window.setTimeout(function () {
+                    // if Mojit gets loading twice, then Object.keys(Mojito.testObjects).length should be 0.
```
Comment (kingo55) [inline tests/test_suite.html]: `if Mojit gets loading twice` -> `if Mojito loads twice`

**Review by kingo55 — COMMENTED** [review]
@allmywant - just a few comments in here to address / discuss.

File: gulpfile.js (hunk 2 above)
Comment (allmywant) [reply gulpfile.js]: @kingo55 
Renamed to `insertDuplicateContainerCheck`.

File: config.js (hunk 3 above)
Comment (allmywant) [reply config.js:12]: @kingo55 
added.

File: tests/test_suite.html (hunk 4 above)
Comment (allmywant) [reply tests/test_suite.html]: @kingo55 
Fixed.

File: tests/test_suite.html [hunk 5]
```diff
@@ -938,7 +938,7 @@
             po.src = '../dist/assets/js/mojito.js';
             po.onload = function () {
                 window.setTimeout(function () {
-                    // if Mojit gets loading twice, then Object.keys(Mojito.testObjects).length should be 0.
+                    // iif Mojito loads twice, then Object.keys(Mojito.testObjects).length should be 0.
```
Comment (kingo55) [inline tests/test_suite.html]: @allmywant - sorry to be pedantic but it looks like there's an extra `i` in here now.

Comment (allmywant) [reply tests/test_suite.html]: @kingo55 
Fixed :)

**Pull request merged**